src/libaktualizr/primary/aktualizr_test.cc

```diff
@@ -492,6 +492,355 @@ TEST(Aktualizr, FullWithUpdatesNeedReboot) {
+    ASSERT_TRUE(installed_versions[0].IsValid());
+    ASSERT_EQ(pending_version, 0);
+    ASSERT_EQ(current_version, 0);
+
+    current_version = 0;  // why it's not boolean
+    pending_version = 0;  // why it's not boolean
+
+    ASSERT_TRUE(
+        storage->loadInstalledVersions(secondary_ecu_id, &installed_versions, &current_version, &pending_version));
+    ASSERT_EQ(installed_versions.size(), 1);
+    ASSERT_TRUE(installed_versions[0].IsValid());
+    ASSERT_EQ(pending_version, 0);
+    ASSERT_EQ(current_version, 0);
+  }
+}
+
+/*
+ * Initialize -> UptaneCycle -> download updates -> fail installation
+ *
+ * Verifies whether the uptane client is not at pending state after installation failure
+ *
+ * Checks actions:
+ *
+ * - [x] Update is not in pending state anymore after failed installation
+ * - [x] Store negative device installation result when an ECU installation failed
+ */
+TEST(Aktualizr, InstallationFailure) {
+  TemporaryDirectory temp_dir;
+  auto http_server_mock = std::make_shared<HttpInstallationFailed>(temp_dir.Path());
+  Config conf = UptaneTestCommon::makeTestConfig(temp_dir, http_server_mock->tls_server);
+  conf.uptane.polling_sec = 0;
+  conf.bootloader.reboot_sentinel_dir = temp_dir.Path();
+  auto storage = INvStorage::newStorage(conf.storage);
+
+  std::vector<std::string> expected_event_order = {
+      "UpdateCheckComplete",    "DownloadProgressReport", "DownloadTargetComplete", "DownloadProgressReport",
+      "DownloadTargetComplete", "AllDownloadsComplete",   "InstallStarted",         "InstallTargetComplete",
+      "InstallStarted",         "InstallTargetComplete",  "AllInstallsComplete",    "PutManifestComplete"};
+
+  std::vector<std::string> expected_report_order = {
+      "EcuDownloadStarted",     "EcuDownloadCompleted",     "EcuDownloadStarted",     "EcuDownloadCompleted",
+      "EcuInstallationStarted", "EcuInstallationCompleted", "EcuInstallationStarted", "EcuInstallationCompleted"};
+
+  const std::string primary_ecu_id = "CA:FE:A6:D2:84:9D";
+  const std::string secondary_ecu_id = "secondary_ecu_serial";
+
+  {
+    fiu_init(0);
+    fiu_enable("fake_package_install", 1, nullptr, 0);
+
+    Aktualizr aktualizr(conf, storage, http_server_mock);
+    EventHandler event_hdlr(aktualizr);
+    aktualizr.Initialize();
+
+    // verify currently installed version
+    std::vector<Uptane::Target> installed_versions;
+    size_t current_version{0};
+    size_t pending_version{0};
+
+    ASSERT_TRUE(
+        storage->loadInstalledVersions(primary_ecu_id, &installed_versions, &current_version, &pending_version));
+
+    // for some reason there is no any installed version at initial Aktualizr boot/run
+    // IMHO it should return currently installed version
+    ASSERT_TRUE(installed_versions.empty());
+    ASSERT_EQ(pending_version, 0);
+    ASSERT_EQ(current_version, 0);
+
+    aktualizr.UptaneCycle();
+    aktualizr.uptane_client_->completeInstall();
```

Comment:
Is it necessary to even try calling `completeInstall` if the installation failed?

Reply:
Yes, if we would like emulate Aktualizr app's behavior we expect on a device. If the installation failed completeInstall will be called from RunForever().